File: backend/app/features/config/normalizers.py
"""
Config feature normalizers — transform raw Aruba API data into UI-ready format.
"""
import ipaddress
from typing import Any, Dict, List, Optional


def normalize_allow_list_client(client: Any) -> Optional[Dict[str, Any]]:
    if not isinstance(client, dict):
        return None

    client_id = str(client.get("clientId") or client.get("macAddress") or "").strip()
    mac_address = str(client.get("macAddress") or client_id).strip()
    client_name = str(client.get("clientName") or mac_address or client_id).strip()
    if not client_id or not mac_address:
        return None

    return {
        "clientId": client_id,
        "macAddress": mac_address,
        "ipAddress": str(client.get("ipAddress") or "").strip(),
        "clientName": client_name or client_id,
        "clientType": str(client.get("clientType") or "").strip(),
        "wiredNetworkIds": [str(value) for value in client.get("wiredNetworkIds", []) if value not in (None, "")] if isinstance(client.get("wiredNetworkIds"), list) else [],
        "connectionDurationInSeconds": int(client.get("connectionDurationInSeconds") or 0),
        "isWatchlisted": bool(client.get("isWatchlisted", False)),
    }


def normalize_allow_list(allow_list: Any) -> Dict[str, Any]:
    if not isinstance(allow_list, dict):
        return {
            "id": "",
            "allowListState": None,
            "maxEntityAllowedAllowLists": None,
            "isAllowListEnabled": False,
            "maxAllowedClients": None,
            "isExtendAllowListEnabled": False,
            "allowedClients": [],
            "availableClients": [],
        }

    allowed_clients = [
        normalized
        for normalized in (
            normalize_allow_list_client(client)
            for client in allow_list.get("allowedClients", [])
        )
        if normalized
    ] if isinstance(allow_list.get("allowedClients"), list) else []

    available_clients = [
        normalized
        for normalized in (
            normalize_allow_list_client(client)
            for client in allow_list.get("availableClients", [])
        )
        if normalized
    ] if isinstance(allow_list.get("availableClients"), list) else []

    return {
        "id": str(allow_list.get("id") or "").strip(),
        "allowListState": allow_list.get("allowListState"),
        "maxEntityAllowedAllowLists": allow_list.get("maxEntityAllowedAllowLists"),
        "isAllowListEnabled": bool(allow_list.get("isAllowListEnabled", False)),
        "maxAllowedClients": allow_list.get("maxAllowedClients"),
        "isExtendAllowListEnabled": bool(allow_list.get("isExtendAllowListEnabled", False)),
        "allowedClients": allowed_clients,
        "availableClients": available_clients,
    }


def find_wireless_network_in_wired_response(
    wired_response: Dict[str, Any],
    network_id: str,
) -> Optional[Dict[str, Any]]:
    for wired_network in wired_response.get("elements", []):
        if not isinstance(wired_network, dict):
            continue
        for wireless_network in wired_network.get("wirelessNetworks", []) or []:
            if not isinstance(wireless_network, dict):
                continue
            if str(wireless_network.get("id") or wireless_network.get("networkId") or "") == str(network_id):
                return wireless_network
    return None


def normalize_allowed_destinations(destinations: Any) -> List[str]:
    if not isinstance(destinations, list):
        return []

    normalized: List[str] = []
    for destination in destinations:
        value = ""
        if isinstance(destination, str):
            value = destination.strip()
        elif isinstance(destination, dict):
            value = str(
                destination.get("address")
                or destination.get("ipAddress")
                or destination.get("value")
                or ""
            ).strip()
        elif destination not in (None, ""):
            value = str(destination).strip()

        if value and value not in normalized:
            normalized.append(value)

    return normalized


def normalize_access_point(access_point: Any) -> Optional[Dict[str, Any]]:
    if isinstance(access_point, dict):
        ap_id = str(
            access_point.get("deviceId")
            or access_point.get("id")
            or access_point.get("serialNumber")
            or access_point.get("macAddress")
            or access_point.get("name")
            or ""
        )
        ap_name = (
            access_point.get("deviceName")
            or access_point.get("name")
            or access_point.get("hostname")
            or access_point.get("serialNumber")
            or ap_id
        )
        if not ap_id and not ap_name:
            return None

        enabled_radio_bands = access_point.get("enabledRadioBands")
        return {
            "id": ap_id or ap_name,
            "name": ap_name or ap_id,
            "deviceId": ap_id or ap_name,
            "deviceName": ap_name or ap_id,
            "deviceModel": access_point.get("deviceModel", ""),
            "radioBandMapping": access_point.get("radioBandMapping", ""),
            "enabledRadioBands": enabled_radio_bands if isinstance(enabled_radio_bands, list) else [],
            "isBoundToNetwork": bool(access_point.get("isBoundToNetwork", True)),
        }

    if access_point in (None, ""):
        return None

    value = str(access_point)
    return {
        "id": value,
        "name": value,
        "deviceId": value,
        "deviceName": value,
        "deviceModel": "",
        "radioBandMapping": "",
        "enabledRadioBands": [],
        "isBoundToNetwork": True,
    }


def normalize_individual_wireless_network(
    network: Dict[str, Any],
    allow_list: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    authentication = network.get("authentication", "psk")
    security = str(network.get("security", "wpa3")).lower()
    security_option = "wpa23_personal"
    if authentication == "psk" and security == "wpa2":
        security_option = "wpa2_personal"
    elif authentication == "eap" and security == "wpa2":
        security_option = "wpa2_enterprise"
    elif authentication == "eap" and security == "wpa3":
        security_option = "wpa23_enterprise"

    access_points = []
    for ap in network.get("accessPoints") or []:
        normalized_ap = normalize_access_point(ap)
        if normalized_ap:
            access_points.append(normalized_ap)

    is_access_restricted = bool(network.get("isAccessRestricted", False))
    is_specific_destinations_allowed = bool(
        is_access_restricted and network.get("isSpecificDestinationsAllowed", False)
    )
    allowed_destinations = (
        normalize_allowed_destinations(network.get("allowedDestinations"))
        if is_specific_destinations_allowed
        else []
    )
    normalized_allow_list = normalize_allow_list(
        allow_list if isinstance(allow_list, dict) else network.get("allowList")
    )

    return {
        "id": str(network.get("networkId") or network.get("id") or ""),
        "networkKind": "wireless",
        "displayName": network.get("networkName", "Unnamed SSID"),
        "label": f"{network.get('networkName', 'Unnamed SSID')} · Wireless",
        "networkName": network.get("networkName", "Unnamed SSID"),
        "isEnabled": network.get("isEnabled", True),
        "health": network.get("health", "good"),
        "state": "Active" if network.get("isEnabled", True) else "Disabled",
        "usage": network.get("type", "employee"),
        "authentication": authentication,
        "security": security,
        "securityOption": security_option,
        "wiredNetworkId": network.get("wiredNetworkId"),
        "ipAddressingMode": network.get("ipAddressingMode", "internal"),
        "dhcpScope": network.get("dhcpScope") if isinstance(network.get("dhcpScope"), dict) else {},
        "isAvailableOn24GHzRadioBand": network.get("isAvailableOn24GHzRadioBand", False),
        "isAvailableOn5GHzRadioBand": network.get("isAvailableOn5GHzRadioBand", False),
        "isAvailableOn6GHzRadioBand": network.get("isAvailableOn6GHzRadioBand", False),
        "isLegacy80211bRatesEnabled": network.get("isLegacy80211bRatesEnabled", False),
        "isSsidHidden": network.get("isSsidHidden", False),
        "preSharedKey": network.get("preSharedKey", ""),
        "isAccessRestricted": is_access_restricted,
        "isInternetAllowed": True,
        "isSpecificDestinationsAllowed": is_specific_destinations_allowed,
        "allowedDestinations": allowed_destinations,
        "allowList": normalized_allow_list,
        "accessPoints": access_points,
    }


def normalize_individual_wired_network(network: Dict[str, Any]) -> Dict[str, Any]:
    is_enabled = (
        network.get("networkActivationEffectivePolicy", {}).get("isActive")
        if isinstance(network.get("networkActivationEffectivePolicy"), dict)
        else None
    )
    if is_enabled is None:
        is_enabled = network.get("isEnabled", True)

    display_name = network.get("wiredNetworkName") or f"VLAN {network.get('vlanId', 'Unknown')}"
    return {
        "id": str(network.get("id") or network.get("networkId") or network.get("vlanId") or ""),
        "networkKind": "wired",
        "displayName": display_name,
        "label": f"{display_name} · Wired",
        "networkName": display_name,
        "isEnabled": is_enabled,
        "health": network.get("health", "good"),
        "state": "Active" if is_enabled else "Disabled",
        "usage": network.get("type", "employee"),
        "vlanId": network.get("vlanId"),
        "isIgmpSnoopingEnabled": bool(network.get("isIgmpSnoopingEnabled", False)),
        "isAccessRestricted": bool(network.get("isAccessRestricted", False)),
        "allowedDestinations": normalize_allowed_destinations(network.get("allowedDestinations")),
        "isDhcpArpProtectionEnabled": bool(
            network.get(
                "shouldApplyNetworkSecurityProtections",
                network.get("isDhcpArpProtectionEnabled", False),
            )
        ),
    }

File: frontend/src/pages/SiteConfig/normalizers.js
/**
 * SiteConfig normalizers — pure functions for transforming network API data into UI state.
 * Reusable across SiteConfig sections and potentially other modules.
 */
import { SECURITY_OPTIONS } from './constants';

export const resolveSecurityOption = (network) =>
    SECURITY_OPTIONS.find(
        (option) => option.authentication === (network?.authentication || 'psk') && option.security === String(network?.security || 'wpa2').toLowerCase()
    )?.id || 'wpa23_personal';

export const normalizeAccessPoints = (accessPoints) => (
    Array.isArray(accessPoints)
        ? accessPoints
            .map((accessPoint) => {
                if (accessPoint && typeof accessPoint === 'object') {
                    const id = String(
                        accessPoint.id
                        || accessPoint.deviceId
                        || accessPoint.serialNumber
                        || accessPoint.macAddress
                        || accessPoint.name
                        || ''
                    );
                    const name = String(
                        accessPoint.name
                        || accessPoint.deviceName
                        || accessPoint.hostname
                        || id
                    );

                    if (!id && !name) return null;

                    return {
                        id: id || name,
                        name: name || id,
                        deviceId: String(accessPoint.deviceId || id || name),
                        deviceName: String(accessPoint.deviceName || name || id),
                        deviceModel: String(accessPoint.deviceModel || ''),
                        radioBandMapping: String(accessPoint.radioBandMapping || ''),
                        enabledRadioBands: Array.isArray(accessPoint.enabledRadioBands) ? accessPoint.enabledRadioBands : [],
                        isBoundToNetwork: Boolean(accessPoint.isBoundToNetwork),
                    };
                }

                if (accessPoint == null || accessPoint === '') return null;

                const value = String(accessPoint);
                return {
                    id: value,
                    name: value,
                    deviceId: value,
                    deviceName: value,
                    deviceModel: '',
                    radioBandMapping: '',
                    enabledRadioBands: [],
                    isBoundToNetwork: true,
                };
            })
            .filter(Boolean)
        : []
);

export const formatAccessPointBands = (accessPoint) => {
    const bands = Array.isArray(accessPoint?.enabledRadioBands) ? accessPoint.enabledRadioBands : [];
    if (bands.length > 0) {
        const labelMap = {
            '2.4ghz': '2.4 GHz',
            '5ghz': '5 GHz',
            '6ghz': '6 GHz',
        };
        return bands.map((band) => labelMap[String(band).toLowerCase()] || String(band)).join(', ');
    }

    const mappingMap = {
        '2.4ghz_and_5ghz': '2.4 GHz, 5 GHz',
        '2.4ghz_only': '2.4 GHz',
        '5ghz_only': '5 GHz',
        '6ghz_only': '6 GHz',
        '5ghz_and_6ghz': '5 GHz, 6 GHz',
        '2.4ghz_5ghz_and_6ghz': '2.4 GHz, 5 GHz, 6 GHz',
    };
    return mappingMap[String(accessPoint?.radioBandMapping || '').toLowerCase()] || '';
};

export const buildWirelessForm = (network) => ({
    mode: 'wireless',
    networkName: network?.networkName || '',
    isEnabled: network?.isEnabled !== false,
    state: network?.state || 'Active',
    wiredNetworkId: network?.wiredNetworkId || '',
    ipAddressingMode: network?.ipAddressingMode || 'internal',
    networkAddress: network?.dhcpScope?.network || '',
    subnetMask: network?.dhcpScope?.netmask || '255.255.255.0',
    dnsServerAssignationMode: network?.dhcpScope?.dns?.dnsServerAssignationMode || 'automatic',
    primaryDnsServer: network?.dhcpScope?.dns?.primaryDnsServer || '',
    secondaryDnsServer: network?.dhcpScope?.dns?.secondaryDnsServer || '',
    isAvailableOn24GHzRadioBand: Boolean(network?.isAvailableOn24GHzRadioBand),
    isAvailableOn5GHzRadioBand: Boolean(network?.isAvailableOn5GHzRadioBand),
    isAvailableOn6GHzRadioBand: Boolean(network?.isAvailableOn6GHzRadioBand),
    isLegacy80211bRatesEnabled: Boolean(network?.isLegacy80211bRatesEnabled),
    isSsidHidden: Boolean(network?.isSsidHidden),
    preSharedKey: network?.preSharedKey || '',
    health: network?.health || 'good',
    type: network?.usage || 'employee',
    authentication: network?.authentication || 'psk',
    security: String(network?.security || 'wpa3').toLowerCase(),
    securityOption: network?.securityOption || resolveSecurityOption(network),
    accessPoints: normalizeAccessPoints(network?.accessPoints),
});

export const buildWiredForm = (network) => ({
    mode: 'wired',
    networkName: network?.networkName || `VLAN ${network?.vlanId ?? ''}`.trim(),
    isEnabled: network?.isEnabled !== false,
    state: network?.state || 'Active',
    health: network?.health || 'good',
    type: network?.usage || 'employee',
    vlanId: network?.vlanId ?? '',
    isIgmpSnoopingEnabled: Boolean(network?.isIgmpSnoopingEnabled),
    isDhcpArpProtectionEnabled: Boolean(network?.isDhcpArpProtectionEnabled),
    isNetworkDestinationsRestricted: Boolean(
        network?.isAccessRestricted
        || network?.isNetworkDestinationsRestricted
        || (Array.isArray(network?.allowedDestinations) && network.allowedDestinations.length > 0)
    ),
    allowedDestinationIpAddresses: Array.isArray(network?.allowedDestinations)
        ? network.allowedDestinations
            .map((destination) => {
                if (typeof destination === 'string') return destination;
                if (destination && typeof destination === 'object') {
                    return String(destination.address || destination.ipAddress || destination.value || '').trim();
                }
                return '';
            })
            .filter(Boolean)
        : [],
});

export const buildFormForNetwork = (network) => (
    network?.networkKind === 'wired' ? buildWiredForm(network) : buildWirelessForm(network)
);

export const getAllowListClientKey = (client) =>
    String(client?.macAddress || client?.clientId || '').trim().toLowerCase();

export const normalizeAllowListClient = (client) => {
    if (!client || typeof client !== 'object') return null;

    const macAddress = String(client.macAddress || '').trim();
    const clientId = String(client.clientId || macAddress).trim();
    const clientName = String(client.clientName || macAddress || clientId).trim();
    if (!macAddress || !clientId) return null;

    return {
        clientId,
        macAddress,
        ipAddress: String(client.ipAddress || '').trim(),
        clientName: clientName || macAddress,
        clientType: String(client.clientType || '').trim(),
        wiredNetworkIds: Array.isArray(client.wiredNetworkIds)
            ? client.wiredNetworkIds.map((value) => String(value || '').trim()).filter(Boolean)
            : [],
        connectionDurationInSeconds: Number(client.connectionDurationInSeconds || 0),
        isWatchlisted: Boolean(client.isWatchlisted),
    };
};

export const normalizeAllowListClients = (clients) => (
    Array.isArray(clients)
        ? clients.map((client) => normalizeAllowListClient(client)).filter(Boolean)
        : []
);

export const buildWirelessAllowListState = (network) => {
    const allowList = network?.allowList && typeof network.allowList === 'object' ? network.allowList : {};
    const allowedClients = normalizeAllowListClients(allowList.allowedClients);
    const availableClients = normalizeAllowListClients(allowList.availableClients);
    const isAllowListEnabled = Boolean(
        allowList.isAllowListEnabled || allowList.isExtendAllowListEnabled
    );

    return {
        allowListId: String(allowList.id || '').trim(),
        allowListState: allowList.allowListState || null,
        maxEntityAllowedAllowLists: allowList.maxEntityAllowedAllowLists ?? null,
        maxAllowedClients: allowList.maxAllowedClients ?? null,
        isAllowListEnabled,
        isExtendAllowListEnabled: Boolean(allowList.isExtendAllowListEnabled),
        allowedClients,
        availableClients,
    };
};

export const buildWirelessAccessControlState = (network) => {
    const allowListState = buildWirelessAllowListState(network);
    return {
        isNetworkDestinationsEnabled: Boolean(
            network?.isAccessRestricted
            || network?.isSpecificDestinationsAllowed
            || (Array.isArray(network?.allowedDestinations) && network.allowedDestinations.length > 0)
        ),
        isSpecificClientsEnabled: allowListState.isAllowListEnabled,
        isInternetAllowed: true,
        isSpecificIpAddressEnabled: Boolean(
            network?.isSpecificDestinationsAllowed
            || (Array.isArray(network?.allowedDestinations) && network.allowedDestinations.length > 0)
        ),
        allowedDestinationIpAddresses: Array.isArray(network?.allowedDestinations)
            ? network.allowedDestinations
                .map((destination) => {
                    if (typeof destination === 'string') return destination;
                    if (destination && typeof destination === 'object') {
                        return String(destination.address || destination.ipAddress || destination.value || '').trim();
                    }
                    return '';
                })
                .filter(Boolean)
            : [],
        ...allowListState,
    };
};

export const mergeAllowListIntoAccessControlState = (currentState, allowList) => ({
    ...currentState,
    ...buildWirelessAllowListState({ allowList }),
});

export const mergeSpecificClientSelectionIntoAccessControlState = (currentState, selectedClients, allowList = {}) => {
    const normalizedSelectedClients = normalizeAllowListClients(selectedClients);
    const hasAllowListUpdate = allowList && typeof allowList === 'object' && Object.keys(allowList).length > 0;
    const normalizedAllowList = hasAllowListUpdate ? buildWirelessAllowListState({ allowList }) : {};
    const nextAllowedClientsByKey = new Map();

    normalizeAllowListClients([
        ...(hasAllowListUpdate ? normalizedAllowList.allowedClients : []),
        ...currentState?.allowedClients,
        ...normalizedSelectedClients,
    ]).forEach((client) => {
        nextAllowedClientsByKey.set(getAllowListClientKey(client), client);
    });

    const nextAllowedClients = Array.from(nextAllowedClientsByKey.values());
    const allowedClientKeys = new Set(nextAllowedClients.map((client) => getAllowListClientKey(client)));
    const nextAvailableClients = normalizeAllowListClients([
        ...(hasAllowListUpdate ? normalizedAllowList.availableClients : []),
        ...currentState?.availableClients,
    ]).filter((client) => !allowedClientKeys.has(getAllowListClientKey(client)));

    return {
        ...currentState,
        ...normalizedAllowList,
        isSpecificClientsEnabled: true,
        isAllowListEnabled: true,
        allowedClients: nextAllowedClients,
        availableClients: nextAvailableClients,
    };
};

export const buildNetworkWithAllowList = (network, allowList, networkName) => {
    if (!network) return network;

    return {
        ...network,
        networkName: networkName || network.networkName,
        displayName: networkName || network.displayName,
        label: `${networkName || network.networkName || network.displayName || 'Unnamed SSID'} · Wireless`,
        allowList: {
            ...(network.allowList && typeof network.allowList === 'object' ? network.allowList : {}),
            ...(allowList && typeof allowList === 'object' ? allowList : {}),
            allowedClients: normalizeAllowListClients(allowList?.allowedClients),
            availableClients: normalizeAllowListClients(allowList?.availableClients),
        },
    };
};

export const buildAccessControlSignature = (state) => JSON.stringify({
    isNetworkDestinationsEnabled: Boolean(state?.isNetworkDestinationsEnabled),
    isSpecificClientsEnabled: Boolean(state?.isSpecificClientsEnabled),
    isSpecificIpAddressEnabled: Boolean(state?.isSpecificIpAddressEnabled),
    allowedDestinationIpAddresses: Array.isArray(state?.allowedDestinationIpAddresses)
        ? [...state.allowedDestinationIpAddresses].sort()
        : [],
    allowedClients: normalizeAllowListClients(state?.allowedClients)
        .map((client) => ({
            clientId: client.clientId,
            macAddress: client.macAddress.toLowerCase(),
        }))
        .sort((left, right) => left.macAddress.localeCompare(right.macAddress)),
});

export const isWirelessAccessControlDirty = (state, network) => (
    buildAccessControlSignature(state) !== buildAccessControlSignature(buildWirelessAccessControlState(network))
);

export const isValidIpv4Address = (value) => {
    if (!value) return false;
    const parts = value.trim().split('.');
    if (parts.length !== 4) return false;
    return parts.every((part) => {
        if (!/^\d+$/.test(part)) return false;
        const number = Number(part);
        return number >= 0 && number <= 255;
    });
};

export const getNetworkKey = (network) => {
    const networkKind = String(network?.networkKind || '');
    const networkId = String(network?.id || '');
    return networkKind && networkId ? `${networkKind}:${networkId}` : networkId;
};

export const findInitialNetwork = (networks, requestedNetworkKey, requestedNetworkId) => (
    networks.find((network) => getNetworkKey(network) === requestedNetworkKey) ||
    networks.find((network) => String(network.id) === String(requestedNetworkId)) ||
    networks[0] ||
    null
);

export const normalizeInitialConfigurationNetwork = (network) => {
    if (!network || !network.networkKind || !network.id) return null;

    const displayName = network.displayName || network.networkName || `VLAN ${network.vlanId ?? '—'}`;
    const usage = network.usage || network.type || 'employee';
    const state = network.state || (network.isEnabled === false ? 'Disabled' : 'Active');

    if (network.networkKind === 'wired') {
        return {
            ...network,
            id: String(network.id),
            networkKind: 'wired',
            displayName,
            label: network.label || `${displayName} · Wired`,
            networkName: network.networkName || displayName,
            isEnabled: network.isEnabled !== false,
            health: network.health || 'good',
            state,
            usage,
            vlanId: network.vlanId ?? '',
            isIgmpSnoopingEnabled: Boolean(network.isIgmpSnoopingEnabled),
            isDhcpArpProtectionEnabled: Boolean(network.isDhcpArpProtectionEnabled),
            isNetworkDestinationsRestricted: Boolean(
                network.isAccessRestricted
                || network.isNetworkDestinationsRestricted
                || (Array.isArray(network.allowedDestinations) && network.allowedDestinations.length > 0)
            ),
            allowedDestinations: Array.isArray(network.allowedDestinations) ? network.allowedDestinations : [],
        };
    }

    return {
        ...network,
        id: String(network.id),
        networkKind: 'wireless',
        displayName,
        label: network.label || `${displayName} · Wireless`,
        networkName: network.networkName || displayName,
        isEnabled: network.isEnabled !== false,
        health: network.health || 'good',
        state,
        usage,
        authentication: network.authentication || 'psk',
        security: String(network.security || 'wpa3').toLowerCase(),
        wiredNetworkId: network.wiredNetworkId || '',
        ipAddressingMode: network.ipAddressingMode || 'internal',
        dhcpScope: typeof network.dhcpScope === 'object' && network.dhcpScope !== null ? network.dhcpScope : {},
        isAvailableOn24GHzRadioBand: Boolean(network.isAvailableOn24GHzRadioBand),
        isAvailableOn5GHzRadioBand: Boolean(network.isAvailableOn5GHzRadioBand),
        isAvailableOn6GHzRadioBand: Boolean(network.isAvailableOn6GHzRadioBand),
        isLegacy80211bRatesEnabled: Boolean(network.isLegacy80211bRatesEnabled),
        isSsidHidden: Boolean(network.isSsidHidden),
        preSharedKey: network.preSharedKey || '',
        accessPoints: normalizeAccessPoints(network.accessPoints),
    };
};
